File: interpreter-for-cpp/src/Runtime/Executer.cpp
#include "Executer.h"
#include "../Grammar/ParseTool.h"
#include "Sentence/Sentence.h"
#include "Space.h"

using namespace peak::interpreter;

std::shared_ptr<Executer> Executer::Create(const std::string& src) {
	if (src.empty()) {
		return nullptr;
	}
	auto parseData = ParseTool::Load(src);
	if (!parseData->bSuccess) {
		return nullptr;
	}
	return std::shared_ptr<Executer>(new Executer(parseData));
}

Executer::Executer(std::shared_ptr<ParseData> data)
	: _parseData(data) {
	_space = std::shared_ptr<Space>(new Space(SpaceType::None));
}

Executer::~Executer() {
	_space->Clear();
}

bool Executer::Execute() {
	_space->Clear();
	for (auto sentence : _parseData->sentenceList) {
		if (!Sentence::IsSuccess(sentence->Execute(_space))) {
			return false;
		}
	}
	return true;
}

std::shared_ptr<Space> Executer::GetSpace() const {
	return _space;
}

std::shared_ptr<Variable> Executer::FindVariable(const std::string& name) {
	return _space->FindVariable(name);
}
bool Executer::AddVariable(std::shared_ptr<Variable> variable) {
	return _space->AddVariable(variable);
}

File: interpreter-for-cs/src/Runtime/Executer.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

namespace peak.interpreter {

	public class Executer {
		private ParseData _parseData = null;
		private Space _space = null;
		public Space space { get => _space; }

		public static Executer Create(string src) {
			if (string.IsNullOrEmpty(src)) {
				return null;
			}
			var parseData = ParseTool.Load(src);
			if (!parseData.bSuccess) {
				return null;
			}
			return new Executer(parseData);
		}
		private Executer(ParseData data) {
			_parseData = data;
			_space = new Space(SpaceType.None);
		}
		~Executer() {
			_space.Clear();
		}

		public bool Execute() {
			_space.Clear();
			foreach (var sentence in _parseData.sentenceList) {
				if (!Sentence.IsSuccess(sentence.Execute(_space))) {
					return false;
				}
			}
			return true;
		}

		public Variable FindVariable(string name) {
			return _space.FindVariable(name);
		}
		public bool AddVariable(Variable variable) {
			return _space.AddVariable(variable);
		}

	}

} // namespace peak.interpreter